### DISCUSSION TITLE
Control Flow in Git

### DISCUSSION THREAD

**Discussion by Frizk1429** [community/community#142471] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

How to manipulate the flow of commits across branches using Git rebase, particularly in a scenario where you want to merge multiple feature branches in a specific order while keeping the commit history linear?

**Comment by fxrdhan** [comment, marked as answer]
Hi there!

To manipulate commit flows across branches using Git rebase while maintaining a linear history when merging multiple feature branches, here's a detailed approach:

## Basic Approach
```bash
# Assuming we have feature branches: feature-A, feature-B, and feature-C
# And we want to merge them in that specific order

# First, ensure you're up to date with main
git checkout main
git pull origin main

# Start with feature-A
git checkout feature-A
git rebase main
git checkout main
git merge feature-A

# Move to feature-B
git checkout feature-B
git rebase main
git checkout main
git merge feature-B

# Finally, feature-C
git checkout feature-C
git rebase main
git checkout main
git merge feature-C
```

## Advanced Scenarios
### 1. When features are interdependent:
```bash
# If feature-B depends on feature-A
git checkout feature-B
git rebase feature-A

# If feature-C depends on both A and B
git checkout feature-C
git rebase feature-B
```

### 2. Cherry-picking specific commits:
```bash
# To select specific commits from a feature branch
git checkout main
git cherry-pick <commit-hash>
```

### 3. Interactive rebase for commit reorganization:
```bash
# To reorganize commits in a feature branch
git checkout feature-branch
git rebase -i main

# This opens an editor where you can:
# - reorder commits by changing their line order
# - squash commits using 'squash' or 's'
# - edit commit messages using 'reword' or 'r'
# - drop commits using 'drop' or 'd'
```

## Best Practices
### 1. *Always backup branches* before complex rebase operations:
```bash
git branch feature-A-backup feature-A
```

### 2. *Use `--onto` for advanced rebasing:*
```bash
# Rebasing a range of commits onto a new base
git rebase --onto new-base old-base feature-branch
```

### *3. Resolve conflicts systematically:*
* When conflicts occur during rebase, resolve them in each commit step by step
* Use `git rebase --continue` after resolving each conflict
* If needed, use `git rebase --abort` to start over

## Important Considerations
Never rebase published branches unless team agreements permit it.\
Ensure all team members are aware of the rebase strategy.\
Document the branch merging order in your project guidelines.\
Use meaningful commit messages that clearly indicate feature relationships.

## Visual Guide to the Process
```plaintext
Before:
main     A---B---C
feature-A    D---E
feature-B        F---G
feature-C            H---I

After rebase:
main     A---B---C---D---E---F---G---H---I
```

**Reply by Frizk1429** [reply]
thanks bro 👍
